rfcs/convergence/simplify-prop-merging.md

```diff
@@ -0,0 +1,326 @@
+(baseState as unknown) as ExtendedComponentProps).slot = undefined;
+const state = mergeProps(baseState, defaultProps, resolveShorthandProps(props, extendedComponentShorthandProps));
+```
+
+Both of the examples above are quite opinionated and require knowledge of the internals of `mergeProps` and `resolveShorthands` to understand without guidance.
+
+Ideally conformance tests would be required to validate defaultProps always goes through `resolveShorthandProps`.
+
+In the case where a component is extended by adding slots, it's pretty hard to validate both in terms of functionality and type safety.
+
+### Merging functions and properties with primitives
+
+Functions or primitives used in `mergeProps` will simply replace the value in the order of application in. This is extremely similar to native object spreading but without a deep merge functionality.
+
+Prop objects tend to be relatively flat, and in cases where objects/arrays are used that need to be deep merged on each render we should examine those cases and make exceptions.
+
+## Detailed Design or Proposal
+
+### Declare `as` and the rendered `component` separately
+
+```tsx
+const state = {
+  fooSlot: { type: "submit" },
+  barSlot: { as: "button" },
+  components: {
+    fooSlot: "button",
+    barSlot: Loader, // JSX element
+  },
+};
+
+// let's use our `Component`
+<Component fooSlot={{ as: 'div' }} barSlot={{as: "div"}} />
+
+// result
+<Component>
+  <div type="submit">button</div>
+  <Loader as="div"} />
+</Component>
+```
+
+The `as` prop is now a prop of the component being rendered in the slot and no longer defines what is rendered in the slot.
+
+### Do not support `as` prop for slots that render native DOM elements
+
+In the previous section, it should be noted that `fooSlot` does not use the `as` prop since it is an element that renders a native HTML element. We propose to enforce this.
+
+Enforcing this requirement will give type safety to our native HTML slots, and make sure consumers don't unknowingly dangerously spread DOM attributes.
+
+It is still possible to render different DOM elements, but this now has to be done explicitly without any type safety. It's also not easy to provide type safety easily in this case currently in Typescript.
+
+```tsx
+// still typed to button props
+// not recommended ❌
+// `components` should not even by exposed as a prop
+<Component components={{ fooSlot: 'div' }} foo={{}} />;
+
+// Can be done by creating component state variants
+const state = useComponent();
+state.components.fooSlot = 'div';
+
+// We have similar utilities
+// Safe attributes for the native slot
+// Can be done but still not ideal
+state.fooSlot = filterNativeDivProps(state.fooSlot);
+
+// Recommended way, let the consumer control everything 🚀
+// We return what would originally be rendered in the children render function
+<Component
+  fooSlot={{
+    children: (Component, props) => {
+      /** Can be a rocket 🚀 or a footgun 🔫 */
+    },
+  }}
+/>;
+```
```

Comment:
````suggestion
```tsx
// ✅ Recommended way, let the consumer control everything 🚀
// We return what would originally be rendered in the children render function
<Component
  fooSlot={{
    children: (Component /* a real element type, for example "div" */, props /* a real type for props */) => {
      /** Can be a rocket 🚀 or a footgun 🔫 */
    },
  }}
/>;
```
````